### ISSUE TITLE
Adding ImageProcessor.Web.Config to use remote images breaks web.config

### ISSUE THREAD

**Issue by Blatant** [JimBobSquarePants/ImageProcessor#150]
Hi JB,

Sorry to bother you but I've hit a pretty difficult issue for me! 

I just installed ImageProcessor and ImageProcessor.Web via nuGet (which caused Umbraco to upgrade, whcih caused issues, don't get me started! lol). Anyway I got these in place no problem.

I want to handle remote images so after building and testing (obviously the remote image did not resize) I looked around and realised I needed to add ImageProcessor.Web.Config. I installed this using nuGet, no problem, cleaned, emptied cache, built and ran. now I get a YSOD with this info : 

Server Error in '/' Application.
Configuration Error
Description: An error occurred during the processing of a configuration file required to service this request. Please review the specific error details below and modify your configuration file appropriately.

Parser Error Message: An error occurred creating the configuration section handler for imageProcessor/processing: Could not load type 'ImageProcessor.Web.Configuration.ImageProcessingSection' from assembly 'ImageProcessor.Web'.

Source Error:

Line 18:   < sectionGroup name="imageProcessor">
Line 19:       < section name="security" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageSecuritySection, ImageProcessor.Web" />
Line 20:       < section name="processing" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageProcessingSection, ImageProcessor.Web" />
Line 21:       < section name="caching" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageCacheSection, ImageProcessor.Web" />
Line 22:     < /sectionGroup>< /configSections>

Source File: C:\Users\Dan\Documents\Visual Studio 2013\Projects\EnergieComplete4\EnergieUmbrella2Project\web.config    Line: 20 

I've tried managing my web.config however I just can't tell what exactly is out of place! Any ASAP help would be much appriciated!

PS> Versions installed are IP 1.9.5.0, IP.Web 4.2.1.0 IP.Web.Config 2.2.0.0

**Comment by Blatant** [comment]
Hi JB,

I have progressed the issue a bit, it seems that the web.config section : 

      <sectionGroup name="imageProcessor">
          <section name="security" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageSecuritySection, ImageProcessor.Web" />
          <section name="processing" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageProcessingSection, ImageProcessor.Web" />
          <section name="caching" requirePermission="false" type="ImageProcessor.Web.Configuration.ImageCacheSection, ImageProcessor.Web" />
      </sectionGroup>

Should in fact be : 

      <sectionGroup name="imageProcessor">
          <section name="security" requirePermission="false" type="ImageProcessor.Web.Config.ImageSecuritySection, ImageProcessor.Web" />
          <section name="processing" requirePermission="false" type="ImageProcessor.Web.Config.ImageProcessingSection, ImageProcessor.Web" />
          <section name="caching" requirePermission="false" type="ImageProcessor.Web.Config.ImageCacheSection, ImageProcessor.Web" />
      </sectionGroup>

I checked the namespaces for this. However the next error is as so : 

Server Error in '/' Application.
Configuration Error
Description: An error occurred during the processing of a configuration file required to service this request. Please review the specific error details below and modify your configuration file appropriately.

Parser Error Message: Unrecognized attribute 'preserveExifMetaData'. Note that attribute names are case-sensitive.

Source Error:


Line 1:  <?xml version="1.0" encoding="utf-8"?>
Line 2:  <processing preserveExifMetaData="false">
Line 3:    <presets>
Line 4:    </presets>


Source File: C:\Users\Dan\Documents\Visual Studio 2013\Projects\EnergieComplete4\EnergieUmbrella2Project\config\imageprocessor\processing.config    Line: 2 


Am I suffering from version issues or something?? I installed all latest from nuGet???

**Comment by Blatant** [comment]
a little further here, I went ahead and mashed up the configs myself.

in processing.config jsut removing the offending attribute works.

The next error was it didn't like the elements in security.config. I used the legacy (pre 4.1) version you briefly snippeted here : http://imageprocessor.org/imageprocessor-web/configuration/#securityconfig

and I can now proceed. no YSOD. I would like to understand wahts going on here though as a basic Nuget Install is not playing fair for me!

Not actually getting a remote resized image, but I think I maty have syntax wrong (going to try legacy here too)

Out of curiosity, as there's no "real" examples in the documentation assuming the following what should my src attribute be exactly : 

* The image path is at http://my.cdn.com/some/path/image.png
* The URL http://my.cdn.com/ is added to security.config
* I wish to resize the image to 200px x 200px
* I wish to crop the image

**Comment by Blatant** [comment]
Last comment I (might) promise!

Switching to the legacy remote url syntax gets me there.

Can you possibly help me work out what version conflicts I'm having? I installed from nuGet without specifying version (so latest) but it just didn't install the corrilating things from the looks of it!

**Comment by JimBobSquarePants** [comment]
Looks like Nuget is messing you around. If you install the latest versions each then they should work.

As you can see on Nuget your versioning is all wrong. You haven't got the latest versions.

https://www.nuget.org/packages/ImageProcessor.Web.Config/

ImageProcessor.Web.Config 2.2.0.0 requires:
 - ImageProcessor (≥ 2.2.0.0)
 - ImageProcessor.Web (≥ 4.2.0.0)

**You shouldn't be using the old config syntax if you have the latest version at all.**
Configuration details can be found here: http://imageprocessor.org/imageprocessor-web/configuration/

Remote file syntax is indicated on the documentation as follows:
http://imageprocessor.org/imageprocessor-web/imageprocessingmodule/#remote

```
<!-- Recommended syntax as of version 4.2 -->
/remote.axd/your-external-image?width=300
```

That would make your url
```
/remote.axd/my.cdn.com/some/path/image.png?width=200&crop=200&mode=crop

```

You cannot suffix querystring parameters onto a remote url without the correct prefix as there is no way to intercept an outgoing call.

The example syntax for resize can be found in the documentation here.
http://imageprocessor.org/imageprocessor-web/imageprocessingmodule/resize/

Hopefully that should clear things up...

I would like to know why updating ImageProcessor would cause Umbraco to update? 

ImageProcessor cannot cause that to happen since they are separate libraries. What I am assuming you meant was that an older version of Umbraco had ImagePorcessor as a dependency but with a maximum allowable version set?

**Comment by Blatant** [comment]
Hi JB, 

Thanks for the response, I agree I seem to have got conflicting versions. Though I now have it working so things should be simpler this time! Here's an accurate script on what happened, it does seem like something is messed up within nuGet : 

1. Went to http://imageprocessor.org/ and clicked the ImageProcessor link (not being sure which pieces I needed) to NuGet. Typed the install command in : 

        PM> Install-Package ImageProcessor
2. This is where the fun happened,This was version 1.9.5.0. I admit I don't understand the intricacies of nuGet but it seems that NuGet auto ran the Umbraco update, either like you say with a maximum version dependancy or perhaps nuGet auto ran update on other Nuget libs (I installed U7 with NuGet too) the PM log stated it was uninstalling among others Umbraco 7.1.4.

3. Then it re-installed Umbraco and the other libraries, all automated. Sure enougth nuGet warned about overwriting all the configs ect, and when I restarted the site I got the upgrade screens. Finished the update and re-applied my configs (connection settings, custom AppSettings, ect).

4. So I got Umbraco back up and installed ImageProcessor.Web (again from the nuget page). IP.Web depends on IP 2.2.0 so IP got updated at this point too.

         PM> Install-Package ImageProcessor.Web
5. This gave me resizing of images local to the web server, but not remote (could be I had my paths wrong). Understandable, a bit more reading and I found I needed ImageProcessor.Web.Config anyway for setting the whitelist so

        PM> Install-Package ImageProcessor.Web.Config

6. This is where the problems in my next posts cropped up. I applied the fixes in my next posts and got it working. By now I had latest versions (IP 2.2.0.0, IP.Web 4.2.1.0, IP.Web.Config 2.2.0.0). This is where I am at right now! I just double checked the installed versions! However here are my configs : 

config/imageprocessor/security.config

    <?xml version="1.0" encoding="utf-8"?>
    <security allowRemoteDownloads="true" timeout="300000" maxBytes="4194304" remotePrefix="/remote.axd">
      
      <whiteList>
        <add url="http://my.contentdistrib.net"/>
      </whiteList>
    </security>

config/imageprocessor/processing.config (truncated, just showing removed attribute on `<processing>`

    <?xml version="1.0" encoding="utf-8"?>
    <processing>
      <presets>
      </presets>

Web.config (imagePrecessor nodes extracted)

    <configuration>
        <configSections>
            <sectionGroup name="imageProcessor">
                <section name="security" requirePermission="false" type="ImageProcessor.Web.Config.ImageSecuritySection, ImageProcessor.Web" />
                <section name="processing" requirePermission="false" type="ImageProcessor.Web.Config.ImageProcessingSection, ImageProcessor.Web" />
                <section name="caching" requirePermission="false" type="ImageProcessor.Web.Config.ImageCacheSection, ImageProcessor.Web" />
            </sectionGroup>
        </configSections>
        
        <imageProcessor>
            <security configSource="config\imageprocessor\security.config" />
            <caching configSource="config\imageprocessor\cache.config" />
            <processing configSource="config\imageprocessor\processing.config" />
        </imageProcessor>
    <configuration>

So, here we are, new style configs arn't accepted but old versions are, all versions seem up to date! 

** Remote URL syntax **
I have to admit the only reason I am using the old version of syntax here is that it's obvious (from the http://) that you include the domain, path and image name. You answer my question here with your example, thanks!

**Comment by JimBobSquarePants** [comment]
Ok... Do you actually have the correct versions installed? If your references are not pointing to the latest version numbers then you do not. 

Nuget simply doesn't work in the way you described either I'm afraid... You must have told it to update all your references. I'mageProcessor has no visibility of Umbraco, none of your package.config files will list Umbraco as a a dependency so wiithout updating all your references the behaviour you are describing is impossible.

In the old syntax the double `?` was used to get ImageProcessor.Web to recognise the requested url and the querystring as separate objects. The new syntax does not do that unless your original url also requires querystring parameters. (Static Google Maps Api for example).

I would uninstall all the packages and start again, installing ImageProcessor.Web.Config first so that the correct dependencies are definitely installed.

From then on use the new syntax...

**Comment by Blatant** [comment]
Hi JB,

I think I've gotten to the bottom of this now! Thanks for all your support it is much appriciated! 

What has happened is that I had origionally installed an older version of ImageProcessor at the beginning of development, subsiquently I've gotten my local enviroment into package version hell!

In short NuGet did its job, updating the IP package and its dependant packages when I installed IP.Web and IP.Web.Config. NuGet did a "check for updates" thing during my tinkering which downloaded Umbraco7.1.9 package, I agree with you that it's nothing to do with your NuGet packages!

However as I mention I had prevoiusly installed IP, which I did not realise. I had existing references within my project, which needed updating to reference the upgraded packages once I had sorted out these versions.

Thanks for your assistance JB, implimentation got the better of me on this one!
